lib/libzfs/libzfs_diff.c

```diff
@@ -597,7 +597,7 @@ get_snapshot_names(differ_info_t *di, const char *fromsnap,
 	tsnlen = strlen(tosnap) - tdslen;	/* includes @ sign */
 
 	if (fsnlen <= 1 || tsnlen == 1 || (fdslen == 0 && tdslen == 0) ||
-	    (fsnlen == 0 && tsnlen == 0)) {
+	    (fdslen == 0 && tsnlen == 0)) {
```

Comment:
This isn't at all clear.  But I believe simply removing the extra condition is the correct fix.  The code here is attempting to determine the needed dataset names and snapshot names when partial information is provided.

```diff
diff --git a/lib/libzfs/libzfs_diff.c b/lib/libzfs/libzfs_diff.c
index 419189a..972cfed 100644
--- a/lib/libzfs/libzfs_diff.c
+++ b/lib/libzfs/libzfs_diff.c
@@ -596,8 +596,7 @@ get_snapshot_names(differ_info_t *di, const char *fromsnap,
        fsnlen = strlen(fromsnap) - fdslen;     /* includes @ sign */
        tsnlen = strlen(tosnap) - tdslen;       /* includes @ sign */
 
-       if (fsnlen <= 1 || tsnlen == 1 || (fdslen == 0 && tdslen == 0) ||
-           (fsnlen == 0 && tsnlen == 0)) {
+       if (fsnlen <= 1 || tsnlen == 1 || (fdslen == 0 && tdslen == 0)) {
                return (zfs_error(hdl, EZFS_INVALIDNAME, di->errbuf));
        } else if ((fdslen > 0 && tdslen > 0) &&
            ((tdslen != fdslen || strncmp(fromsnap, tosnap, fdslen) != 0))) {
```